compose-backstack/src/main/java/com/zachklipp/compose/backstack/ChildSavedStateRegistry.kt

```diff
@@ -1,69 +1,91 @@
 package com.zachklipp.compose.backstack
 
 import androidx.compose.Composable
+import androidx.compose.CompositionLifecycleObserver
+import androidx.compose.currentComposer
 import androidx.compose.remember
 import androidx.ui.savedinstancestate.UiSavedStateRegistry
 import androidx.ui.savedinstancestate.UiSavedStateRegistryAmbient
-import androidx.ui.savedinstancestate.rememberSavedInstanceState
 
 /**
  * Returns a [UiSavedStateRegistry] that will automatically save values from all its registered
  * providers whenever [childWillBeComposed] transitions from true to false, and make those values available
  * to be restored when [childWillBeComposed] transitions from false to true.
  */
 @Composable
-@Suppress("RemoveExplicitTypeArguments")
 fun ChildSavedStateRegistry(childWillBeComposed: Boolean): UiSavedStateRegistry {
-    val parentRegistry = UiSavedStateRegistryAmbient.current
-
-    // This map holds all the savedInstanceState for this screen as long as it exists
-    // in the backstack. When the screen is hidden, we will cache its state providers
-    // into this map before removing it from the composition. This cache will in turn
-    // be persisted into and restored from the parent UiSavedStateRegistry.
-    val values = rememberSavedInstanceState<MutableMap<String, Any>> { mutableMapOf() }
-    val holder = remember {
-        // If there's no registry available, then we won't be restored anyway so there are no
-        // serializability restrictions on saved values.
-        val canBeSaved: (Any) -> Boolean = parentRegistry?.let { it::canBeSaved } ?: { true }
-        SavedStateHolder(canBeSaved, values)
-    }
-    holder.setScreenVisibility(childWillBeComposed)
-    return holder.registry
+    val parent = UiSavedStateRegistryAmbient.current
+    val key = currentComposer.currentCompoundKeyHash.toString()
+    val holder = remember { SavedStateHolder() }
+    return holder.updateAndReturnRegistry(key, parent, childWillBeComposed)
 }
 
-internal class SavedStateHolder(
-    private val canBeSaved: (Any) -> Boolean,
-    private var values: Map<String, Any>
-) {
-    var registry: UiSavedStateRegistry = createRegistry()
-        private set
+internal class SavedStateHolder : CompositionLifecycleObserver {
+    private var key: String? = null
```

Comment:
Actually, I'm pretty sure this can be passed into the constructor and will never change. That would simplify this code a bit and remove the null assertions below.

Reply:
Yeah I think you're right, it differs from `rememberSavedInstanceState` because that function allows for custom keys.